### PR TITLE
Use automasking in seq2seq example

### PR DESCRIPTION
**Pull request by marcvanzee**
Use automasking to create ragged batches for the seq2seq example. This ends up simplifying especially the Encoder significantly since we do no longer have to keep track of hidden state of all `eos_id` tokens.

In this PR, I only use automasking during training and evaluation, but we can extend this for decoding. I plan to do this in a follow-up step.

Thanks to @mattjj and @j-towns for making this great technology possible! Especially thanks to @j-towns for guiding me through all the bugs and error messages.

### CODE REVIEW

**Comment by marcvanzee** [comment]
**Note:** the accuracy is not as good as before yet (it only reaches about 70%), so I am debugging the reason why. I expect there is some bug in padding, or in computing the loss, since the loss doesn't go below 0.2 (before it was 0.005). Moreover, decoding never seems to give a correct answer. 

Example output of current run:
```
I0908 20:25:31.590295 140558216381824 train.py:384] train step: 7200, loss: 0.2025, accuracy: 61.72
I0908 20:25:31.604599 140558216381824 train.py:345] DECODE: 81+807 = 870 (INCORRECT) correct=888
I0908 20:25:31.604734 140558216381824 train.py:345] DECODE: 51+98 = 1006 (INCORRECT) correct=149
I0908 20:25:31.604794 140558216381824 train.py:345] DECODE: 50+847 = 878 (INCORRECT) correct=897
I0908 20:25:31.604840 140558216381824 train.py:345] DECODE: 6+856 = 648 (INCORRECT) correct=862
I0908 20:25:31.604883 140558216381824 train.py:345] DECODE: 53+715 = 746 (INCORRECT) correct=768
```

File: linen_examples/seq2seq/train.py [hunk 1]
```diff
@@ -130,12 +130,13 @@ def encode_onehot(batch_inputs, max_len):
...
-    if len(tokens) > max_len:
+    unpadded_len = len(tokens)
+    if unpadded_len > max_len:
       raise ValueError(f'Sequence too long ({len(tokens)}>{max_len}): \'{s}\'')
     tokens = np.pad(tokens, [(0, max_len-len(tokens))], mode='constant')
-    return onehot(tokens, CTABLE.vocab_size)
+    return onehot(tokens, CTABLE.vocab_size), unpadded_len
 
-  return np.array([encode_str(inp) for inp in batch_inputs])
+  return np.array([encode_str(inp) for inp in batch_inputs], dtype=object)
```
Comment (levskaya) [inline linen_examples/seq2seq/train.py:139]: What is the point of wrapping this in a `np.array(..., dtype=object)`?  Don't the below uses work just as well if this were just a list?

File: linen_examples/seq2seq/train.py [hunk 2]
```diff
@@ -352,24 +302,35 @@ def compute_metrics(logits, labels):
   return metrics
 
 
+IN_SHAPES = [{'query': '(n, _)', 'answer': '(m, _)'}]
+OUT_ELEM = f'(m +- 1, {CTABLE.vocab_size})'
```
Comment (levskaya) [inline linen_examples/seq2seq/train.py:306]: curious: what does `m +- 1` syntax mean here?

File: linen_examples/seq2seq/train.py [hunk 3]
```diff
@@ -223,31 +179,26 @@ def __call__(self, carry, x):
...
-    projection = nn.Dense(features=self.vocab_size, name='projection')
-    lstm_state, y = lstm_cell(lstm_state, x)
-    logits = projection(y)
-    predicted_tokens = jax.random.categorical(categorical_rng, logits)
-    prediction = onehot(predicted_tokens, self.vocab_size)
+    lstm_state, y = nn.LSTMCell()(lstm_state, x)
+    logits = nn.Dense(features=CTABLE.vocab_size)(y)
+    predicted_token = jax.random.categorical(categorical_rng, logits)
+    prediction = jnp.array(predicted_token == jnp.arange(CTABLE.vocab_size), 
+                           dtype=jnp.float32)
```
Comment (levskaya) [inline linen_examples/seq2seq/train.py:186]: just curious: why return things in one-hot form instead of just integer ids?  It's probably obvious I'm just dumb about this.

**Review by levskaya — APPROVED** [review]
Everything looks pretty clear to me, just a few points of curiosity.

File: linen_examples/seq2seq/train.py (hunk 1 above)
Comment (marcvanzee) [reply linen_examples/seq2seq/train.py:139]: If I don't add this then I get the following warning:

```
train.py:139: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return np.array([encode_str(inp) for inp in batch_inputs])
```

File: linen_examples/seq2seq/train.py (hunk 2 above)
Comment (marcvanzee) [reply linen_examples/seq2seq/train.py:306]: This is the somewhat funny notation to denote "m-1". The decoder outputs "m-1" tokens since the input also contains the start token "=".

File: linen_examples/seq2seq/train.py (hunk 1 above)
Comment (levskaya) [reply linen_examples/seq2seq/train.py:139]: but why wrap this in an ndarray at all??  why not just keep it as a list?

File: linen_examples/seq2seq/train.py (hunk 2 above)
Comment (levskaya) [reply linen_examples/seq2seq/train.py:306]: ok cool that makes sense.

File: linen_examples/seq2seq/train.py (hunk 3 above)
Comment (marcvanzee) [reply linen_examples/seq2seq/train.py:186]: I think the whole logic of outputting both logits and predictions is a bit funny. I think it would be easier to only output the logits, and then use them to get the predictions, if necessary. However, in this PR I tried to keep the logic mostly the same except for the automasking, so I think that would be nice as a follow-up. WDYT?

File: linen_examples/seq2seq/train.py (hunk 2 above)
Comment (andsteing) [reply linen_examples/seq2seq/train.py:306]: Minor nit : Could we make it `m + -1` then? Then it's clearer that we add `-1` as opposed to inventing a new operator `+-` with mysterious semantics :)

File: linen_examples/seq2seq/train.py (hunk 3 above)
Comment (levskaya) [reply linen_examples/seq2seq/train.py:186]: Oh sure, I was just curious about it!

Comment (marcvanzee) [reply linen_examples/seq2seq/train.py:186]: No problem! I agree with your observation :-)

File: linen_examples/seq2seq/train.py (hunk 1 above)
Comment (marcvanzee) [reply linen_examples/seq2seq/train.py:139]: Ahh yes that actually works! Nice, changed it.

File: linen_examples/seq2seq/train.py (hunk 2 above)
Comment (marcvanzee) [reply linen_examples/seq2seq/train.py]: Sure, good idea! Done.
